test/python/circuit/test_controlled_gate.py

```diff
@@ -1021,12 +1021,13 @@ def test_standard_base_gate_setting(self, gate_class):
         """
         if gate_class in {SingletonControlledGate, _SingletonControlledGateOverrides}:
             self.skipTest("SingletonControlledGate isn't directly instantiated.")
-        num_free_params = len(_get_free_params(gate_class.__init__, ignore=["self"]))
+        gate_params = _get_free_params(gate_class.__init__, ignore=["self"])
+        num_free_params = len(gate_params)
         free_params = [0.1 * i for i in range(num_free_params)]
-        if gate_class in [MCU1Gate, MCPhaseGate]:
-            free_params[1] = 3
-        elif gate_class in [MCXGate]:
-            free_params[0] = 3
+        # set number of control qubits
+        for i in range(num_free_params):
+            if gate_params[i] == "num_ctrl_qubits":
+                free_params[i] = 3
```

Comment:
I understand that this change is to make the code a bit general and not only work for `MCU1Gate`, `MCPhaseGate` and `MCXGate`, but I was wondering, why did this fail with the new Aer?

Reply:
The test discovery attempts to discover every registered subclass of `ControlledGate` (Python makes that information easily accessible), without restriction to the Qiskit standard library, where it would make sense.

We ideally should probably fix the test discovery to limit it only to controlled gates defined in `qiskit.circuit.library.standard_gates`.

Reply:
So you are suggesting changing the current code on line 1017:
```@data(*ControlledGate.__subclasses__())```
with something like: 
```@data([cls for cls in allGates.__dict__.values() if isinstance(cls,type) and issubclass(cls, ControlledGate)])```?

Reply:
I'm not a huge fan of the automatic discovery in its current form at all, but given the current file, yeah, something like that would be fine.  Something a little less magic would be
```python
from qiskit.circuit.library.standard_gates import get_standard_gate_name_mapping

standard_controlled_types = [x.base_class for x in get_standard_gate_name_mapping().values() if isinstance(x, ControlledGate)]
```

That would exclude the multi-controlled gates, which imo should be handled specially if they're not going to follow the conventions of standard-library gates.

(_really_, I don't think the mcx etc gates should be in `standard_gates` at all, they should be in `generalized_gates`, but that's an argument for another time)

Reply:
Given that it looks like there is a lot of room for improvement in these tests, let's keep the current fix as-is and consider improving the tests in a follow-up.